Migration symlinks

mdegerne opened the pull request.

timuralp commented:
I haven't started on the PR yet, but we should add tests for the missing coverage bits:
- shunt: https://coveralls.io/builds/23359920/source?filename=s3_sync%2Fshunt.py#L300 (integration tests might cover this, but it wouldn't hurt to add unit tests)
- migrator: https://coveralls.io/builds/23359920/source?filename=s3_sync%2Fmigrator.py#L951

timuralp commented on s3_sync/migrator.py:
```diff
@@ -919,13 +923,80 @@ def object_queue_put(
+                           SYMLINK_OBJ, SYMLINK_ACCT]:
+                    pass
+                else:
+                    headers[k] = v
```
Is there any other field we can set here that's not already enumerated? What else do listings contain?

mdegerne replied:
eliminated this branch

timuralp commented on s3_sync/migrator.py:
```diff
@@ -919,13 +923,80 @@ def object_queue_put(
+            target = headers[SYMLINK_OBJ]
+            acct_target = headers.get(SYMLINK_ACCT)
+        else:
+            headers = {'content_length': '0'}
```
Can symlink objects not have their own arbitrary user metadata? We never do a HEAD to look it up, which means the assumption is they can't have own metadata, but I wasn't sure if that's indeed the limitation on them. Given the code above for the `symlink=get` case, I assume they actually can have own metadata?

mdegerne replied:
Dang it! I took a comment from Tim as gospel when I should have checked. You are correct. The link can have it's own user metadata.

mdegerne replied:
fixed

timuralp commented on s3_sync/migrator.py:
```diff
@@ -919,13 +923,80 @@ def object_queue_put(
+        work = None
+        try:
+            symlink_acct, symlink_target = symlink_target_from_listing(remote)
+            if symlink_target:
```
What if the `symlink_target` doesn't exist, but exception wasn't raised? I suppose that should never happen, but we should probably program defensively for that. Should it fall through into the regular work in that case?

mdegerne replied:
Implementing alternate (easier to read) logic

timuralp commented on s3_sync/migrator.py:
```diff
@@ -919,13 +923,80 @@ def object_queue_put(
         else:
             self.verify_queue.put(work)
 
+    def _create_symlink(self, work):
```
We could also do:
```
def _create_symlink(self, container, key, target, acct_target, remote_container, timestamp, entry):
```
and expand the arguments when calling this method.

mdegerne replied:
ok

timuralp commented on s3_sync/migrator.py:
```diff
@@ -919,13 +923,80 @@ def object_queue_put(
+
+        params = dict(symlink='put')
+        with self.ic_pool.item() as ic:
+            try:
```
This seems like copy/paste from `_upload_object`. Can we call `_upload_object`? Is the only thing missing the `params` argument? I'd change that method to not expect a tuple, but rather explicit arguments. Most of the places we call it, we are creating a tuple just to pass it to `_upload_object` method (seems a little silly).

mdegerne replied:
ok. fixed

timuralp commented on s3_sync/shunt.py:
```diff
@@ -278,7 +278,55 @@ def __call__(self, env, start_response):
+        if req.method not in ['GET', 'HEAD']:
+            return self.app(env, start_response)
+
+        # Try action
```
Nit: I'd consider replacing the 4 ensuing comments with a single comment description of why we're doing this.

mdegerne replied:
Yeah, I just left the comments from my outline in. I am going back to fix this now.

timuralp commented on s3_sync/shunt.py:
```diff
@@ -278,7 +278,55 @@ def __call__(self, env, start_response):
+
+        # Try action
+        status, headers, app_iter = req.call_application(self.app)
+        # check if 404
```
Nit: for example this comment is fairly obvious given that it's shortly followed by `if not status.startswith('404 ')`.

timuralp commented on s3_sync/shunt.py:
```diff
@@ -278,7 +278,55 @@ def __call__(self, env, start_response):
+        cont = path_bits[3]
+        obj = '/'.join(path_bits[4:])
+
+        sync_profile = next((self.sync_profiles[(acct, c)]
```
Maybe we should have a small helper that does this. Don't have a strong preference as it'd be pretty small.

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -72,6 +73,17 @@ def __init__(self, *args, **kwargs):
+    @property
+    def remote_account(self):
+        if not self._remote_account:
+            with self.client_pool.get_client() as conn:
```
When is the account different from `self.settings['aws_identity']`? Can we use either that or `self.settings['remote_account']` here?

mdegerne replied:
The problem is that `self.settings['aws_identity']` doesn't always give the <acct> part of the storage url - keystone can result in something completely different, for example.

timuralp replied:
Good point -- keystone can definitely map it to anything else. I'm hesitant about this change because looking up this property is potentially time consuming/blocking. Can we set this as part of any other requests (i.e. in `_call_swift()`)? Do we ever look this up _before_ making an external request? At least maybe we shouldn't decorate it as `property` so as not to give a hint that it's a static/cheap attribute.

timuralp replied:
Talked this over -- we actually make the auth call when the connections are instantiated. While we lazy instantiate the connections, the first call to using one will end up making the auth call, so it's not as big of a concern.

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -216,6 +228,44 @@ def upload_object(self, row, internal_client, upload_stats_cb=None):
             stats_cb=upload_stats_cb)
 
+    def has_symlinks(self):
+        if self.capabilities is None:
```
Consider using `{}` instead of `None`:
```
    if not self.capabilities:
        try:
            self.capabilities = swift_client.get_capabilities()
        except swiftclient.exceptions.ClientException:
            pass    # it will remain an empty dictionary
    return 'symlink' in self.capabilities
```

mdegerne replied:
Taking this in advisement for next patch. For now, code is not part of this patch.

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -216,6 +228,44 @@ def upload_object(self, row, internal_client, upload_stats_cb=None):
+                try:
+                    self.capabilities = swift_client.get_capabilities()
+                except swiftclient.exceptions.ClientException:
+                    self.capabilities = None
```
Assignment wouldn't fail with `ClientException`, so we don't need to assign the value here. We should consider logging the error, though. Probably a warning?

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -216,6 +228,44 @@ def upload_object(self, row, internal_client, upload_stats_cb=None):
+            return 'symlink' in self.capabilities
+        return False
+
+    def create_symlink(self, key, headers, t_obj, t_cont=None, t_acct=None,
```
Please expand `t_*` to `target_*` -- I think it really helps readability.

mdegerne replied:
fixing (and removing to other patch set)

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -216,6 +228,44 @@ def upload_object(self, row, internal_client, upload_stats_cb=None):
+            'put_object', bucket, key, contents='', headers=hdrs,
+            query_string='symlink=put')
+
+    def add_symlink(self, row, internal_client):
```
I don't think anything calls this method? Can we remove it?

mdegerne replied:
removing to other patch set

timuralp commented on s3_sync/utils.py:
```diff
@@ -88,6 +89,29 @@
+    return acct, target
+
+
+def symlink_target_from_row(row):
```
This is only called in `SyncSwift.add_symlink`, which is never called.

mdegerne replied:
removing to other patch set

timuralp replied:
@mdegerne I still this function here? Were you going to move it out?

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -216,6 +228,44 @@ def upload_object(self, row, internal_client, upload_stats_cb=None):
+            return 'symlink' in self.capabilities
+        return False
+
+    def create_symlink(self, key, headers, t_obj, t_cont=None, t_acct=None,
```
There are no callers of `provider.create_symlink` in this patch. Let's at least move it to another patchset where it is used.

mdegerne replied:
removing to other patch set

timuralp commented on test/integration/test_migrator.py:
```diff
@@ -1821,6 +1821,155 @@ def test_s3_large_file_migration(self):
+        conn_noshunt = self.conn_for_acct_noshunt(migration['account'])
+
+        conn_source.put_object(
+            migration['aws_bucket'], 'mylink', '', query_string='symlink=put',
```
As a cheap way to ferret out some of the non-ASCII regressions, let's use non-ASCII characters in the object names.

mdegerne commented on s3_sync/utils.py:
```diff
@@ -88,6 +89,29 @@
+        return None, None
+    parts = path.split('/')
+    if len(parts) >= 5:
+        # Check for duplicate target bug #NNNNNN
```
Added upstream bug reference (#1829085). For reference, what happens is that the symlink_path gets reported as `/v1/<container>/<object>/<container>/<object>` rather than `/v1/<account>/<container>/<object>`

mdegerne commented on s3_sync/utils.py:
```diff
@@ -88,6 +89,29 @@
+        cont = parts[3]
+        if acct in parts[4:] and cont in parts[3:]:
+            # Almost certainily bad information, do head to get real data
+            raise SymlinkPathFailure(path)
```
As above, we don't have the account.

mdegerne commented on s3_sync/migrator.py:
```diff
@@ -919,13 +923,80 @@ def object_queue_put(
+            headers = {'content_length': '0'}
+            for k, v in remote.iteritems():
+                if k == 'hash':
+                    headers['etag'] = v
```
eliminated, this branch is not used

thiagodasilva commented on s3_sync/migrator.py:
```diff
@@ -919,13 +923,80 @@ def object_queue_put(
+                    container, key))
+            src_meta = resp.headers
+            headers = src_meta
+            target = headers[SYMLINK_OBJ]
```
should be headers.get(SYMLINK_OBJ) here too. object might in fact not be a symlink

mdegerne replied:
Hmm. I hadn't thought of this possibility, but there could definitely be a lag and a chance for it to change. I'll code it more defensively.

thiagodasilva commented on s3_sync/migrator.py:
```diff
@@ -919,13 +923,80 @@ def object_queue_put(
+            if resp.status != 200:
+                raise MigrationError('Failed to HEAD "%s/%s"' % (
+                    container, key))
+            src_meta = resp.headers
```
use headers here?

mdegerne replied:
This whole section has been modified.

mdegerne commented on s3_sync/migrator.py:
```diff
@@ -65,6 +67,8 @@
 UploadObjectWork = namedtuple('UploadObjectWork', 'container key object '
                               'headers aws_bucket')
+SymlinkWork = namedtuple('SymlinkWork', 'container key symlink_target '
+                         'symlink_acct aws_bucket timestamp remote')
```
fixed

mdegerne commented on s3_sync/migrator.py:
```diff
@@ -919,13 +923,80 @@ def object_queue_put(
-            aws_bucket, container, remote['name'], timestamp)
+        work = None
+        try:
+            symlink_acct, symlink_target = symlink_target_from_listing(remote)
```
fixed

mdegerne commented on s3_sync/migrator.py:
```diff
@@ -919,13 +923,80 @@ def object_queue_put(
-            aws_bucket, container, remote['name'], timestamp)
+        work = None
+        try:
+            symlink_acct, symlink_target = symlink_target_from_listing(remote)
```
rewriting logic

mdegerne commented on containers/swift-s3-sync/cloud-connector.conf:
```diff
@@ -21,6 +21,9 @@ pipeline = catch_errors gatekeeper healthcheck proxy-logging listing_formats cac
 use = egg:swift-s3-sync#cloud-connector
 conf_file = sync-config.json
 
+[filter: symlink]
```
I think I waffled on this. Fixing.

mdegerne commented on containers/swift-s3-sync/internal-client.conf:
```diff
@@ -1,12 +1,15 @@
 [DEFAULT]
 [pipeline:main]
-pipeline = catch_errors proxy-logging cache proxy-server
+pipeline = catch_errors symlink proxy-logging cache proxy-server
```
Interesting. I saw it left of proxy-logging somewhere. Seems reasonable.

timuralp commented on containers/swift-s3-sync/cloud-connector.conf:
```diff
@@ -13,14 +13,17 @@ eventlet_debug = true
 
 [pipeline:main]
-pipeline = catch_errors gatekeeper healthcheck proxy-logging listing_formats cache swift3 cloud-connector-auth slo proxy-logging proxy-server
+pipeline = catch_errors gatekeeper healthcheck proxy-logging listing_formats cache swift3 cloud-connector-auth slo proxy-logging symlink proxy-server
```
Hm I wonder about this. Do we need symlink to be in the pipeline for cloud connector? I suspect we actually don't, since currently cloud-connector only works with the S3 front end. When adding Swift, it still may not be a feature cloud connector needs in its pipeline.

mdegerne replied:
Yeah. Debated about this internally. I'm happy to remove it if no one thinks it's necessary.

timuralp commented on s3_sync/migrator.py:
```diff
@@ -916,16 +918,71 @@ def _old_enough(self, remote):
+        return self._upload_object(container, key, '', headers,
+                                   remote_container, params)
+
+#        with self.ic_pool.item() as ic:
```
Commented out code. Probably meant to remove it after refactoring?

mdegerne replied:
yep. fixed

timuralp commented on s3_sync/migrator.py:
```diff
@@ -1214,10 +1269,13 @@ def _migrate_slo(self, aws_bucket, slo_container, key, resp, put_headers):
-        container, key, content, headers, aws_bucket = work
+    def _upload_object(self, container, key, content, headers, aws_bucket,
+                       params=None):
+        if params is None:
```
Maybe params should default to {}?

mdegerne replied:
NO. Very bad to do that.
```
>>> def x(foo={}):
...   print foo
...   foo['bar'] = 1
... 
>>> x()
{}
>>> x()
{'bar': 1}
>>> 
```

timuralp replied:
I see, although we never mutate the entity in this case, right? Sure wish we could mark it immutable :(

mdegerne replied:
Yeah. It's a landmine and a danger sign, though, so I never put objects as default options in methods/functions. Tracking down these is a royal pain when they get changed down the road.

thiagodasilva commented on s3_sync/migrator.py:
```diff
@@ -916,16 +918,51 @@ def _old_enough(self, remote):
+    def _create_symlink(self, container, key, remote_container):
+        # were not able to determine target - corrupt listing HEAD link
+        resp = self.provider.head_object(
+            key, container, query_string='symlink=get')
```
should be using remote_container, no?
I ran into a traceback while manually testing, i'm a little concerned that unit/integration tests didn't pick this up, but I haven't looked at tests yet or maybe my testing is wrong

mdegerne replied:
Not sure how that didn't cause an error in the tests. Fixing and seeing about fixing the test side.

thiagodasilva commented on s3_sync/migrator.py:
```diff
@@ -916,16 +918,51 @@ def _old_enough(self, remote):
+        acct_target = headers.get(SYMLINK_ACCT)
+        params = {}
+        if target:
+            if acct_target and acct_target != self.provider.remote_account:
```
I think we should document this behavior

timuralp replied:
Agreed. Could we end up with a symlink to a symlink and a broken link in the end?

thiagodasilva commented on s3_sync/sync_swift.py:
```diff
@@ -260,12 +270,16 @@ def shunt_object(self, req, key):
         if req.method == 'GET':
-            resp = self.get_object(
-                key, resp_chunk_size=65536, headers=headers)
+            args['resp_chunk_size'] = 65536
```
do we need a constant for this value?

mdegerne replied:
We have one. I'll set this to use it.

timuralp replied:
I'd define a new one here (as supposed to using the value from the migrator).

mdegerne commented on containers/swift-s3-sync/cloud-connector.conf:
```diff
@@ -13,14 +13,17 @@ eventlet_debug = true
 
 [pipeline:main]
-pipeline = catch_errors gatekeeper healthcheck proxy-logging listing_formats cache swift3 cloud-connector-auth slo proxy-logging proxy-server
+pipeline = catch_errors gatekeeper healthcheck proxy-logging listing_formats cache swift3 cloud-connector-auth slo proxy-logging symlink proxy-server
```
outdated. But others aren't.

mdegerne commented on containers/swift-s3-sync/proxy-server-noshunt.conf:
```diff
@@ -9,13 +9,16 @@ eventlet_debug = true
 [pipeline:main]
 # Yes, proxy-logging appears twice. This is not a mistake.
-pipeline = healthcheck proxy-logging cache swift3 bulk tempurl versioned_writes slo dlo proxy-logging proxy-server
+pipeline = healthcheck proxy-logging cache swift3 bulk tempurl slo dlo versioned_writes proxy-logging symlink proxy-server
```
fixing

mdegerne commented on containers/swift-s3-sync/proxy-server.conf:
```diff
@@ -16,13 +16,16 @@ name = silver
 [pipeline:main]
 # Yes, proxy-logging appears twice. This is not a mistake.
-pipeline = healthcheck proxy-logging cache swift3 bulk tempurl authtoken keystoneauth tempauth cloud_sync_shunt versioned_writes slo dlo proxy-logging proxy-server
+pipeline = healthcheck proxy-logging cache swift3 bulk tempurl authtoken keystoneauth tempauth cloud_sync_shunt slo dlo versioned_writes proxy-logging symlink proxy-server
```
fixing

thiagodasilva commented on test/integration/test_migrator.py:
```diff
@@ -1821,6 +1821,159 @@ def test_s3_large_file_migration(self):
+
+        headers = {
+            SYMLINK_OBJ: migration['aws_bucket'] + '/test_file',
+            'x-object-meta-testinfo': 'baz'}
```
i'd suggest also setting a custom metadata on the target object and then asserting that after migration the symlink still returns the custom metdata.

thiagodasilva commented on test/integration/test_migrator.py:
```diff
@@ -1821,6 +1821,159 @@ def test_s3_large_file_migration(self):
+                         head_resp[SYMLINK_OBJ])
+        self.assertEqual('baz', head_resp.get('x-object-meta-testinfo'))
+
+    def test_symlink_migrate_other_cont_target(self):
```
what's the expected behavior when the other container is not being migrated? I'd suggest we add tests to "codify" the expected behavior when targeted containers are/aren't being migrated.

thiagodasilva commented on test/integration/test_migrator.py:
```diff
@@ -1821,6 +1821,159 @@ def test_s3_large_file_migration(self):
+            migration['container'], 'mylink', query_string='symlink=get')
+        self.assertEqual('notacontainer/test_file', head_resp[SYMLINK_OBJ])
+
+    def test_symlink_migrate_other_acct_target(self):
```
same question/comment as above

thiagodasilva commented on test/integration/test_migrator.py:
```diff
@@ -1821,6 +1821,159 @@ def test_s3_large_file_migration(self):
+        etag = conn_source.put_object(migration['aws_bucket'], 'link_target',
+                                      'bbbd')
+        conn_destination.put_object(
+            'newcontainer', 'mylink', '', query_string='symlink=put',
```
remove query_string

thiagodasilva commented on test/integration/test_migrator.py:
```diff
@@ -1821,6 +1821,159 @@ def test_s3_large_file_migration(self):
+                         head_resp[SYMLINK_OBJ])
+        self.assertEqual(other_acct, head_resp[SYMLINK_ACCT])
+
+    def test_shunt_symlink_target_only(self):
```
this is really neat behavior!

thiagodasilva commented on test/unit/test_shunt.py:
```diff
@@ -464,6 +464,13 @@ def _test_methods(path, methods=('OPTIONS', 'GET', 'HEAD', 'PUT',
         _test_methods('/v1/AUTH_a/s3/o',
                       ('OPTIONS', 'PUT', 'POST', 'DELETE'))
 
+    def test_symlink_shunt(self):
```
remove?

timuralp replied:
Or rather this should be filled in.

mdegerne commented on s3_sync/migrator.py:
```diff
@@ -916,16 +918,51 @@ def _old_enough(self, remote):
+                headers[SYMLINK_OBJ] = target_container + '/' + target_obj
+                headers[SYMLINK_OBJ] = target
+
+            params['symlink'] = 'put'
```
Well, 💩 ! The fallacy of documentation strikes again. I double checked so many of the bits against the actual code but didn't double check this. Much code to remove.

mdegerne commented on s3_sync/migrator.py:
```diff
@@ -916,16 +918,51 @@ def _old_enough(self, remote):
+                headers[SYMLINK_OBJ] = (
+                    target_container + '/' + target_obj).encode('utf8')
+
+        self.logger.debug('MSD: headers: %s' % (headers,))
```
That's why I use the MSD string, so I don't miss removing them. :P Fixing.

mdegerne commented on s3_sync/migrator.py:
```diff
@@ -916,16 +918,51 @@ def _old_enough(self, remote):
             self.verify_queue.put(work)
 
+    def _create_symlink(self, container, key, remote_container):
+        # were not able to determine target - corrupt listing HEAD link
```
Removing - comment made sense once upon a time.

timuralp commented on s3_sync/utils.py:
```diff
@@ -88,6 +89,33 @@
+SYMLINK_ACCT = 'x-symlink-target-account'
+
+
+def symlink_target_from_listing(headers):
```
@mdegerne did we mean to move this function out?

mdegerne commented:
I thought I did. Fixing.

On Fri, May 24, 2019 at 2:05 PM Timur Alperovich <notifications@github.com>
wrote:

> *@timuralp* commented on this pull request.
> ------------------------------
>
> In s3_sync/utils.py
> <https://github.com/swiftstack/1space/pull/291#discussion_r287514897>:
>
> > +    path = headers.get('symlink_path')
> +    if not path:
> +        return None, None
> +    parts = path.split('/')
> +    if len(parts) >= 5:
> +        # Check for duplicate target bug #NNNNNN
> +        acct = parts[2]
> +        cont = parts[3]
> +        if acct in parts[4:] and cont in parts[3:]:
> +            # Almost certainily bad information, do head to get real data
> +            raise SymlinkPathFailure(path)
> +    _, _ver, acct, target = path.split('/', 3)
> +    return acct, target
> +
> +
> +def symlink_target_from_row(row):
>
> @mdegerne <https://github.com/mdegerne> I still this function here? Were
> you going to move it out?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/swiftstack/1space/pull/291?email_source=notifications&email_token=AAAO3FOHV7AG6753UNEZTGTPXBKBJA5CNFSM4HMTFWOKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZV4RZQ#discussion_r287514897>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAO3FNQX6C32VWE3MNPJVTPXBKBJANCNFSM4HMTFWOA>
> .
>


-- 
Regards,
Mandell Degerness

"True glory consists in doing what deserves to be written; in writing what
deserves to be read; and in so living as to make the world happier for our
living in it."
  Pliny the Elder
